File: src/LinkedList/LC21MergeTwoSortedLists.java
package LinkedList;

public class LC21MergeTwoSortedLists {

    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode head = new ListNode();
        ListNode previous = head;

        while (list1 != null && list2 != null) {
            if (list1.val <= list2.val) {
                previous.next = list1;
                previous = list1;
                list1 = list1.next;
            } else {
                previous.next = list2;
                previous = list2;
                list2 = list2.next;
            }
        }
        previous.next = list1 != null ? list1 : list2;
        return head.next;
    }

    // Definition for singly-linked list.
    public class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }
}

File: src/LinkedList/LC21MergeTwoSortedLists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} list1
 * @param {ListNode} list2
 * @return {ListNode}
 */
var mergeTwoLists = function (list1, list2) {
  let node1 = list1;
  let node2 = list2;
  let curr = new ListNode(0);
  let result = null;
  let count = 0;

  while (node1 || node2) {
    if (!node1) {
      curr.next = node2;
      node2 = node2.next;
    } else if (!node2) {
      curr.next = node1;
      node1 = node1.next;
    } else if (node1.val <= node2.val) {
      curr.next = node1;
      node1 = node1.next;
    } else {
      curr.next = node2;
      node2 = node2.next;
    }
    // store head in result
    if (count == 0) result = curr.next;
    count++;
    curr = curr.next;
  }

  return result;
};
